src/main/java/api/process/FormRecordProcessorHelper.java

```diff
@@ -95,32 +103,46 @@ public static void purgeCases(UserSqlSandbox sandbox) {
-                    "but were missing: " + filter.getMissingCasesString() + ". As a result, the " +
-                    "following cases were also removed from the device: " + filter.getRemovedCasesString());
-        }
+        int removedCaseCount = -1;
+        int removedLedgers = -1;
+        try {
+            sandbox.getConnection().setAutoCommit(false);
+            SqliteIndexedStorageUtility<Case> storage = sandbox.getCaseStorage();
+            DAG<String, int[], String> fullCaseGraph = getFullCaseGraph(storage, new FormplayerCaseIndexTable(sandbox), owners);
+
+            CasePurgeFilter filter = new CasePurgeFilter(fullCaseGraph);
+            if (filter.invalidEdgesWereRemoved()) {
+                Logger.log(LogTypes.SOFT_ASSERT, "An invalid edge was created in the internal " +
+                        "case DAG of a case purge filter, meaning that at least 1 case on the " +
+                        "device had an index into another case that no longer exists on the device");
+                Logger.log(LogTypes.TYPE_ERROR_ASSERTION, "Case lists on the server and device" +
+                        " were out of sync. The following cases were expected to be on the device, " +
+                        "but were missing: " + filter.getMissingCasesString() + ". As a result, the " +
+                        "following cases were also removed from the device: " + filter.getRemovedCasesString());
+            }
 
-        Vector<Integer> casesRemoved = storage.removeAll(filter);
-        removedCaseCount = casesRemoved.size();
-        FormplayerCaseIndexTable indexTable = new FormplayerCaseIndexTable(sandbox);
-        for (int recordId : casesRemoved) {
-            indexTable.clearCaseIndices(recordId);
-        }
+            Vector<Integer> casesRemoved = storage.removeAll(filter.getCasesToRemove());
+            removedCaseCount = casesRemoved.size();
 
+            FormplayerCaseIndexTable indexTable = new FormplayerCaseIndexTable(sandbox);
+            for (int recordId : casesRemoved) {
+                indexTable.clearCaseIndices(recordId);
+            }
 
-        SqliteIndexedStorageUtility<Ledger> stockStorage = sandbox.getLedgerStorage();
-        LedgerPurgeFilter stockFilter = new LedgerPurgeFilter(stockStorage, storage);
-        removedLedgers = stockStorage.removeAll(stockFilter).size();
 
+            SqliteIndexedStorageUtility<Ledger> stockStorage = sandbox.getLedgerStorage();
+            LedgerPurgeFilter stockFilter = new LedgerPurgeFilter(stockStorage, storage);
+            removedLedgers = stockStorage.removeAll(stockFilter).size();
+            sandbox.getConnection().commit();
+        } catch (SQLException e) {
+            e.printStackTrace();
+        } finally {
+            try {
+                sandbox.getConnection().setAutoCommit(true);
+            } catch (SQLException e) {
+                e.printStackTrace();
```

Comment:
Is there a reason this error shouldn't be allowed to bubble up and kill the thread? If this doesn't succeed the sandbox is stuck in an inconsistent state where autocommit isn't enabled, which is a serious issue.